proxy/common/backports.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+"""
+    proxy.py
+    ~~~~~~~~
+    ⚡⚡⚡ Fast, Lightweight, Pluggable, TLS interception capable proxy server focused on
+    Network monitoring, controls & Application development, testing, debugging.
+
+    :copyright: (c) 2013-present by Abhinav Singh and contributors.
+    :license: BSD, see LICENSE for more details.
+"""
+import time
+
+from typing import Any
+
+
+class cached_property:
+    """Decorator for read-only properties evaluated only once within TTL period.
+    It can be used to create a cached property like this::
+
+        import random
+
+        # the class containing the property must be a new-style class
+        class MyClass:
+            # create property whose value is cached for ten minutes
+            @cached_property(ttl=600)
+            def randint(self):
+                # will only be evaluated every 10 min. at maximum.
+                return random.randint(0, 100)
+
+    The value is cached  in the '_cached_properties' attribute of the object instance that
+    has the property getter method wrapped by this decorator. The '_cached_properties'
+    attribute value is a dictionary which has a key for every property of the
+    object which is wrapped by this decorator. Each entry in the cache is
+    created only when the property is accessed for the first time and is a
+    two-element tuple with the last computed property value and the last time
+    it was updated in seconds since the epoch.
+
+    The default time-to-live (TTL) is 300 seconds (5 minutes). Set the TTL to
+    zero for the cached value to never expire.
+
+    To expire a cached property value manually just do::
+        del instance._cached_properties[<property name>]
+
+    Adopted from https://wiki.python.org/moin/PythonDecoratorLibrary#Cached_Properties
+    © 2011 Christopher Arndt, MIT License.
+
+    NOTE: We need this function only because Python in-built are only available
+    for 3.8+.  Hence, we must get rid of this function once proxy.py no longer
+    support version older than 3.8.
+
+    .. spelling::
+
+       backports
```

Comment:
It's unclear why `sphinx` want us to tell it spelling of the filename itself.  We have files named `acceptor`, `executor`, `backport` but still we need to declare them as spellings?  Another redundancy IMHO.

Reply:
PS:  Again this is not known to the authors until they run the `spellcheck-doc` flow.  Nothing from the editor.  I am not sure such `sphinx` extensions exist for VSCode.

Reply:
This is because apidoc generates RST files for each module now. In those files, the document titles contain importable names. And sometimes those names are not in the dictionary. You could somehow modify the templates to wrap the titles with double backticks, I guess. I haven't attempted to solve this because I've spent a lot of effort already and I'm simply exhausted at this point (having a lot more projects to maintain). Sometimes I figure out better ways and then go and update many projects but not for this problem, currently.

As for the editor, it's uncommon to run sphinx from it (or any other command for that matter). Moreover, nobody is supposed to run it directly because you'll end up with people running it with different configurations you don't know anything about and complaining about things that work differently in CI and RTD. Just add a command for running tox if you want.

Also, spellcheck is third-party. You can ask here https://github.com/sphinx-contrib/spelling.

P.S. As I mentioned earlier, many projects just document their API manually, without autogenerating pages for all modules.

Reply:
> I haven't attempted to solve this because I've spent a lot of effort already and I'm simply exhausted at this point (having a lot more projects to maintain)

I totally agree.  You have totally transformed CI/CD/Packaging stuff for `proxy.py`.  Very happy about it.  Really appreciate all the contributions.

> P.S. As I mentioned earlier, many projects just document their API manually, without autogenerating pages for all modules.

But then docs can run out of sync with code and it's a double work to maintain code and keep docs up-to date.  Wdyt about it?

Reply:
> > P.S. As I mentioned earlier, many projects just document their API manually, without autogenerating pages for all modules.
> 
> But then docs can run out of sync with code and it's a double work to maintain code and keep docs up-to date. Wdyt about it?

I need to clarify that with the hybrid approach people often use some of the sphinx directives (`autofunction`?) that may extract parts of the content from docstrings while having accompanying prose in the documents. I think you need to try creating more documents and follow https://diataxis.fr just to have a feeling of how this all works. When you experiment with these things more, maybe it'll become easier to judge where to use prose and what to leave fully autogenerated. Per my observations, the manually crafted parts are usually ones that describe the interconnection between things and the public API while the internal objects/reference may be more generated. I think that in terms of Diataxis, it's reasonable to have separately-maintained Tutorials and How-To Guides while Explanations and References may get more of the generated stuff. Still, it's a balance that you need to find empirically.

Reply:
Cool.  I need to investigate more into this before finalizing on an approach.

My inspiration is coming from Flutter/Dart land, where code contains entire doc site as doc strings.  So as a developer, I hardly have to leave the editor and go to some doc site for reference.  Of-course, flutter team also maintains dedicated tutorial, manuals and showcases for the front page of doc site.

Reply:
Tracking here https://github.com/abhinavsingh/proxy.py/issues/776

Reply:
You still need to have enough information in the docstrings, of course. The developers will be able to call `help()` on objects in the default Python REPL, or `?` in IPython, and various editor integration will be able to extract those. You could also configure Sphinx to generate a man page that would be usable by the downstream consumers so regular Linux distro users would also be able to get that via manpages (`man proxy-py`, for example). But that all doesn't replace the purpose of the types of the docs. I really recommend following the Diataxis principles. Ubuntu is about to adopt it right now and you can also chat with its creator in a dedicated channel on the Write The Docs Slack instance. I've heard that our tech writers at Red Hat have a similar set of guidelines they follow.